### ISSUE TITLE
Stateful logits processors see stale tokens due to lazy evaluation in stream context

### ISSUE THREAD

**Issue by Thump604** [ml-explore/mlx-lm#1156]
## Problem

`generate.py`'s `_step` and MTP `_process_and_sample` pass `tokens`/`prev_tokens` to `logits_processors` while the token arrays are still lazy (unevaluated `mx.concat` results built inside `mx.stream` contexts). Stateful processors like Outlines' FSM-based JSON schema processor read `tokens[-1]` to determine the current state, but get stale prompt tokens instead of the most recently generated token.

## Affected paths

1. **Serial generate `_step`** (line ~427): `tokens = mx.concat([tokens, input_tokens])` followed by `processor(tokens, logits)` — all inside `with mx.stream(generation_stream):`. The concat is lazy; `tokens[-1]` returns the old value.

2. **MTP `_step_backbone`** (line ~594): Same pattern with `prev_tokens`. After `_step_backbone` returns, `prev_tokens` is lazy. `_step_mtp` then passes it to `_process_and_sample` where the processor reads `[-1]` and gets stale data.

3. **BatchGenerator `_step`** (line ~1591): `self.tokens[e]` is a list that grows via `.append()` after `mx.eval` — this path may work correctly since it uses Python lists, not lazy mx.arrays.

## Reproduction

```python
from outlines.backends import get_json_schema_logits_processor
# ... build processor with JSON schema ...
# Pass to stream_generate with logits_processors=[processor]
# Result: "No next state found for the current state: 256 with token ID: 198"
# (198 = newline, the last prompt token, not the generated token)
```

## Suggested fix

Add `mx.eval(tokens)` / `mx.eval(prev_tokens)` before calling logits processors:

```python
# In _step (serial generate):
if logits_processors and len(input_tokens) > 0:
    tokens = mx.concat([tokens, input_tokens]) if tokens is not None else input_tokens
    mx.eval(tokens)  # <-- materialize before stateful processors
    for processor in logits_processors:
        logits = processor(tokens, logits)

# In mtp_generate_step, after _step_backbone:
mx.eval(toks, prev_tokens)  # <-- also materialize prev_tokens
```

And move `_process_and_sample` call outside `with mx.stream()` in `_step_mtp`.

## Impact

This blocks all stateful logits processors (Outlines, guidance, etc.) from working with `stream_generate` and MTP generation. Stateless processors (repetition penalty, etc.) are unaffected since they don't read `tokens[-1]`.

## Environment

- mlx-lm 0.31.2 (vendored)
- mlx 0.27+
- Platform: macOS, Apple Silicon M2 Ultra
- Discovered while integrating Outlines 1.2.12 for guided JSON schema decoding

**Comment by Thump604** [comment]
Additional finding: the `apply_token_bitmask` in `_bias_logits_mlx` correctly returns new masked logits (confirmed: `finite=2` allowed tokens after masking). However, the MTP path calls `_process_and_sample` for BOTH backbone (verified) tokens and draft (speculative) tokens via the same `logits_processors`. 

The issue is that:
1. Backbone call: `prev_tokens` grows, processor advances FSM, returns masked logits ✓
2. MTP draft call: `prev_tokens` is same length (draft doesn't add to it), processor receives the same sequence and tries to re-advance with an already-processed token → FSM error

The processor should either:
- Only be called for backbone (verified) steps, not for draft MTP steps
- Or the draft step should pass the speculative token so the FSM can tentatively advance and rollback on rejection

This is MTP-specific — non-MTP generation with `logits_processors` might work correctly since there's no draft/verify split.

Draft PR #1157 already has the `mx.eval()` materialization fix which is necessary but not sufficient for MTP. The MTP processor integration needs additional logic to handle the draft/verify split.

**Comment by Thump604** [comment]
Additional finding: the masking itself works (confirmed 2 allowed tokens after processor call). The remaining issue is MTP-specific: `_process_and_sample` is called for BOTH backbone (verified) and draft (speculative) steps via the same `logits_processors`. Draft calls see the same `prev_tokens` as the previous backbone call (draft doesn't append to `prev_tokens`), causing the processor to re-process already-handled tokens and error on FSM state advancement. Non-MTP generation may work correctly. PR #1157 addresses the lazy eval materialization; the MTP draft/verify split needs additional processor handling.

**Comment by angeloskath** [comment]
So the original premise is wrong in the sense that an array before and after eval represents the result of the same computation.

Simply put

```python
x = mx.arange(10)
y = mx.array([100])
z = mx.concatenate([x, y])
# z[-1] contains 100
mx.eval(z)
# z[-1] still contains 100
```

So I will close this but if there is a bug feel free to open another one.
